### ISSUE TITLE
bcache & zfs results in minor corruptions

### ISSUE THREAD

**Issue by clefru** [zfsonlinux/zfs#8340]
### System information

Type | Version/Name
 --- | --- 
Distribution Name	| NixOS
Distribution Version	| 18.09.git.a4c4cbb
Linux Kernel	| 4.19.17
Architecture	| x64
ZFS Version	| 0.7.12
SPL Version	| 0.7.12

### Describe the problem you're observing
I have a 2-way mirrored pool, where one mirror leg is involved in an experiment with bcache.

```
zpool
+ mirror
  + LUKS-volume
  | + bcache0
  |   + backing device SATA
  |   + caching device NVME
  + LUKS-volume
    + backing device SATA
```

I set my bcache0 volume to writethrough mode and wait a couple of days to accumulate a bit of data. Then I detach the caching volume essentially making bcache pass through the underlying device. This flushes out the cache to the backing device. Then I issue a "zpool scrub" to see if there was any data corruption in the backing device.

The scrub results are unfortunately never great. I got <10 CKSUM errors on the bache supported leg. None of those CKSUM errors were present in the non-bcache leg.

```
  pool: z
 state: ONLINE
status: One or more devices has experienced an unrecoverable error.  An
        attempt was made to correct the error.  Applications are unaffected.
action: Determine if the device needs to be replaced, and clear the errors
        using 'zpool clear' or replace the device with 'zpool replace'.
   see: http://zfsonlinux.org/msg/ZFS-8000-9P
  scan: scrub repaired 8K in 125h10m with 0 errors on Sat Jan 26 16:44:01 2019
config:

        NAME                                                                                                STATE     READ WRITE CKSUM
        z                                                                                                   ONLINE       0     0     0
          mirror-0                                                                                          ONLINE       0     0     0
            dm-uuid-CRYPT-LUKS1-c7b0e5a217fd4a8080821902c4dee069-LUKS-c7b0e5a2-17fd-4a80-8082-1902c4dee069  ONLINE       0     0     0
            dm-uuid-CRYPT-LUKS1-9448350c7c9c4094831c16f9c2f052a4-LUKS-9448350c-7c9c-4094-831c-16f9c2f052a4  ONLINE       0     0     2

errors: No known data errors
```

I conclude that bcache+ZFS don't play nicely together.

### Describe how to reproduce the problem
I assume that reproducing the hierarchy above should be sufficient. I have no hard VM-reproducible case unfortunately.

**Comment by rincebrain** [comment]
This sounds like a bcache bug report, not a ZFS bug report?

Not that I'm suggesting it can't be a problem with ZoL or a confluence of bugs in each, but if you're using a mirror, one leg with bcache and the other not, and the bcache one turns up with errors, that sounds like something to investigate with bcache?

(I'd also try eliminating the LUKS layer above the bcache disk just to make sure it's not a strange interaction of LUKS and bcache.)

**Comment by richardelling** [comment]
ZFS has a long history of detecting bugs in storage systems. This is a good thing.

The question becomes, how do you root-cause the bug? If bcachefs or the devices 
are not reporting errors (see `dmesg` or system logs) then the next step is to try
and develop a repeatable test. Once it is repeatable, then there are ways to trace
the various paths through the disks to see what is happening.

**Comment by clefru** [comment]
@rincebrain It's unlikely that ZFS is to blame here. OTOH, bcache very carefully designed their "yank-the-power-cord" semantics, so I am not sure if bcache is at fault here. Anyway, the ZFS layer is reporting the bugs, so I found it appropriate to file it here, even if the resolution to this bug is a documentation item that says "Do not use with bcache." (Btw, this is bcache not bcachefs)

@richardelling 
1. Throw it into a VM with a smaller (more scrubable) dataset and cross fingers.
2. Look into the errors on my real world setup more closely. I did have 2 CKSUM errors on this run, and 10 CKSUM errors on the last run. It'd be worth to see if they occur in the same location/data structure and take a wild guess from there. Also the CKSUMs occurred only really late into the scrub, around 99%.

Does somebody know how to dump the error correction that "zpool scrub" does? I'd guess that this is a zfs module parameter, but I can only find,
parm:           zfs_scan_ignore_errors:Ignore errors during resilver/scrub (int)

**Comment by clefru** [comment]
@rincebrain Thank you! I attached the event log here.

Noteworthy maybe:
* `zio_objset = 0x2c` for both errors, so it seems to affect the same kind of object type?
* `time = 0x5c4c7832 0x6186d8b` for both. If that's the creation timestamp this might indicate that one single event caused both cksum failures.
* `zio_blkid = 0x24734` / `zio_blkid = 0x24732` quite close when it comes to corruption location.

Sidenote: I will swap out the NVME module (SM951) and rerun the experiment to rule out a hardware defect. The host is on ECC memory, so that seems okish.

**Comment by rincebrain** [comment]
I ran across #7970 while looking for something else, and since I was thinking what happened here sounds like some cache state not being correctly invalidated, it may be the same type of issue? He also says it's true in 4.18+ but not 4.17, so that'd also be something to test.

**Comment by richardelling** [comment]
In the ereport, the device being repaired is dm-uuid-CRYPT-LUKS1-9448350c7c9c4094831c16f9c2f052a4-LUKS-9448350c-7c9c-4094-831c-16f9c2f052a4
It is not clear from the OP that this is the bcache device. Can you verify?

The object set is not really important here.
The zio_offsets are not contiguous, so it looks more like random failure to me.

**Comment by evolarium** [comment]
@clefru I set up a pool with 4 bcache devices in two mirrors on my home workstation over the weekend.  After copying a couple terabytes to the pool and a couple reboots, I have scrubbed with no errors.

In my case, I needed to modify udev rules to get the bcache backing devices registered in the dracut init.  I also set scsi_mod.scan=sync on the kernel command line because I was initially booting without a zpool.cache, and zfs-import-scan.service would load the pool in a degraded state because only some of the devices would be available when it scanned the default async setting for scsi_mod.  I did see checksum errors when that happened.  I've since setup zpool.cache, and it's been working so far.

**Comment by clefru** [comment]
Sorry for the long response on this and thanks for the responses everyone.

In my bcache setup, I have switched out the caching NVME device for another one to rule out hardware errors. After a long "backin" time, I have scrubbed the new device twice so far and I have not been able to reproduce the corruption. (Granted my kernel and ZFS version have moved on). I will close the issue if my third scrub is successful and blame it on the other NVME module or on the kernel version or on both.
